File: Source/WebGPU/WGSL/Semantics/Types/Texture.cpp
/*
 * Copyright (C) 2022 Apple Inc. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
 * THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "config.h"
#include "Texture.h"

namespace WGSL::Semantics::Types {

Texture::Texture(SourceSpan span, Dimension dimension, UniqueRef<Type> componentType)
    : Type(Type::Kind::Texture, span)
    , m_dimension(dimension)
    , m_componentType(WTFMove(componentType))
{
}

Expected<UniqueRef<Type>, Error> Texture::tryCreate(SourceSpan span, Dimension dimension, UniqueRef<Type> componentType)
{
    if (!componentType->isF32() && !componentType->isI32() && !componentType->isU32()) {
        auto err = makeString(
            "texture component type '",
            componentType->string(),
            "' is not f32, i32 or u32"
        );

        return makeUnexpected(Error { WTFMove(err), componentType->span() });
    }

    // return { makeUniqueRef<Texture>(span, dimension, WTFMove(componentType)) };
    return { UniqueRef(*new Texture(span, dimension, WTFMove(componentType))) };
}

bool Texture::operator==(const Type &other) const
{
    if (!other.isTexture())
        return false;

    const auto& textureOther = downcast<Texture>(other);
    return (m_dimension == textureOther.m_dimension)
        && (m_componentType.get() == textureOther.m_componentType.get());
}

} // namespace WGSL::Semantics::Types

File: Source/WebGPU/Swift Interface/Texture.swift
/*
 * Copyright (c) 2023 Apple Inc. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
 * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import Foundation

public class Texture {
    internal let texture: __WGPUTexture
    private let ownership: Ownership

    internal init(_ texture: __WGPUTexture, ownership: Ownership = .Owned) {
        self.texture = texture
        self.ownership = ownership
    }

    deinit {
        if ownership == .Owned {
            __wgpuTextureRelease(texture)
        }
    }

    public func createView(_ descriptor: WGPUTextureViewDescriptor) -> TextureView {
        var localDescriptor = descriptor
        return TextureView(__wgpuTextureCreateView(texture, &localDescriptor))
    }

    public func destroy() {
        __wgpuTextureDestroy(texture)
    }

    public func getDepthOrArrayLayers() -> UInt32 {
        return __wgpuTextureGetDepthOrArrayLayers(texture)
    }

    public func getDimension() -> WGPUTextureDimension {
        return __wgpuTextureGetDimension(texture)
    }

    public func getFormat() -> WGPUTextureFormat {
        return __wgpuTextureGetFormat(texture)
    }

    public func getHeight() -> UInt32 {
        return __wgpuTextureGetHeight(texture)
    }

    public func getMipLevelCount() -> UInt32 {
        return __wgpuTextureGetMipLevelCount(texture)
    }

    public func getSampleCount() -> UInt32 {
        return __wgpuTextureGetSampleCount(texture)
    }

    public func getUsage() -> WGPUTextureUsage {
        return __wgpuTextureGetUsage(texture)
    }

    public func getWidth() -> UInt32 {
        return __wgpuTextureGetWidth(texture)
    }

    public func setLabel(_ label: String) {
        __wgpuTextureSetLabel(texture, label)
    }
}
